misc/version_updater.py

```diff
@@ -0,0 +1,306 @@
+    errors = []
+
+    for line_nu, line in enumerate(filename.read_text().splitlines()):
+        for regex, expected_line in regexes:
+            res = regex.search(line)
+            if res is not None:
+                matched[regex.pattern] = True
+                matched_part = res.group(0)
+                if matched_part != expected_line:
+                    effective_line = line_nu + 1
+                    effective_col_start = res.start(0)
+                    errors.append(
+                        f"{filename}:{effective_line}:{effective_col_start}: Invalid version in file got `{matched_part}` but expected `{expected_line}`"
+                    )
+
+    match_errors = does_every_regex_where_used(filename, matched)
+    errors.extend(match_errors)
+    return errors
+
+
+def update_tool_version(filename: Path, raw_regexes: RawRegexes, version: str) -> List[str]:
+    try:
+        regexes = compile_regexes(raw_regexes, version)
+    except re.error:
+        raise ValueError(f"Failed to compile regexes for file `{filename}`")
+    matched = {regex[0].pattern: False for regex in regexes}
+
+    with FileInput(filename, inplace=True) as f:
+        for line in f:
+            for regex, replaced_line in regexes:
+                line, substitution_count = regex.subn(replaced_line, line, count=1)
+                if substitution_count >= 1:
+                    matched[regex.pattern] = True
+            print(line, end="")
+
+    return does_every_regex_where_used(filename, matched)
+
+
+def compile_regexes(regexes: RawRegexes, version: str) -> List[Tuple[Pattern[str], str]]:
+    return [regex.compile(version) for regex in regexes]
+
+
+def does_every_regex_where_used(filename: Path, have_matched: Dict[str, bool]) -> List[str]:
+    errors = []
+    for pattern, matched in have_matched.items():
+        if not matched:
+            errors.append(f"{filename}: Regex `{pattern}` found no matches")
+
+    return errors
+
+
+def check_tool(tool: Tool, version: str, update: bool) -> Dict[Path, List[str]]:
+    errors = {}
+
+    check_or_udpate: Callable[[Path, RawRegexes, str], List[str]] = update_tool_version
+    if args.check:
+        check_or_udpate = check_tool_version
+
+    for filename, tools_in_file in FILES_WITH_VERSION_INFO.items():
+        regexes = tools_in_file.get(tool, None)
+        if regexes is None:
+            continue
+
+        files_errors = check_or_udpate(filename, regexes, version)
+
+        if files_errors:
+            errors[filename] = files_errors
+
+    return errors
+
+
+def handle_errors(errors: Dict[Tool, Dict[Path, List[str]]]) -> None:
+    for tool, tool_errors in errors.items():
+        for filename, file_errors in tool_errors.items():
+            print(
+                f"{filename}: Got errors concerning the tool {tool.value} (see below)",
+                file=sys.stderr,
+            )
+            print("\n".join(file_errors), file=sys.stderr)
```

Comment:
```suggestion
            print("\n".join(file_errors), file=sys.stderr)
```

I think the first message adds some noise to the output:
```shell
$ python version_updater.py --check
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml: Got errors concerning the tool rust (see below)
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml:4:2: Invalid version in file got `rust: 1.61.0` but expected `rust: 1.68.0`
/home/marcos/code/parsec-cloud/rust-toolchain.toml: Got errors concerning the tool rust (see below)
/home/marcos/code/parsec-cloud/rust-toolchain.toml:2:0: Invalid version in file got `channel = "1.67.0"` but expected `channel = "1.68.0"`
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml: Got errors concerning the tool node (see below)
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml:5:2: Invalid version in file got `node: 18.2.0` but expected `node: 18.12.0`
```
It would be better to just have the errors:
```shell
$ python version_updater.py --check
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml:4:2: Invalid version in file got `rust: 1.61.0` but expected `rust: 1.68.0`
/home/marcos/code/parsec-cloud/rust-toolchain.toml:2:0: Invalid version in file got `channel = "1.67.0"` but expected `channel = "1.68.0"`
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml:5:2: Invalid version in file got `node: 18.2.0` but expected `node: 18.12.0
```

Even better if is grouped by file (instead of by tool), although that would require to store errors by filename:
```shell
$ python version_updater.py --check
/home/marcos/code/parsec-cloud/rust-toolchain.toml:2:0: Invalid version in file got `channel = "1.67.0"` but expected `channel = "1.68.0"`
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml:4:2: Invalid version in file got `rust: 1.61.0` but expected `rust: 1.68.0`
/home/marcos/code/parsec-cloud/.pre-commit-config.yaml:5:2: Invalid version in file got `node: 18.2.0` but expected `node: 18.12.0
```

Reply:
Having only the actual errors in the output is also better for error counting (If we ever find it useful to display that :)